File: 0001-0500/0101-Symmetric-Tree/py-0101/Solution2.py
# Source : https://leetcode.com/problems/symmetric-tree/
# Author : penpenps
# Time   : 2019-07-09

# Recursively check whether the child node is symmetric or not
# Time Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        # Check whether left and right child are symmetic
        def symmetricEqual(left, right):
            # left and right child are both None, return True
            if not left and not right: return True
            # one of left and right is None, return False
            if not left or not right: return False
            # the value of left and right are not equal, return False
            if left.val != right.val:
                return False
            return symmetricEqual(left.left, right.right) and symmetricEqual(left.right, right.left)
        
        if not root:
            return True

        return symmetricEqual(root.left, root.right)

File: 0001-0500/0098-Validate-Binary-Search-Tree/java-0098/src/Solution2.java
/// Source : https://leetcode.com/problems/validate-binary-search-tree/description/
/// Author : liuyubobobo
/// Time   : 2018-05-22

/// Recursive test
/// Time Complexity: O(n)
/// Space Complexity: O(h)
class Solution2 {

    public boolean isValidBST(TreeNode root) {

        return isValidBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean isValidBST(TreeNode node, int min, int max){

        if(node == null)
            return true;

        if(node.val < min || node.val > max)
            return false;

        if(node.left != null && node.left.val >= node.val)
            return false;

        if(node.right != null && node.right.val <= node.val)
            return false;

        return isValidBST(node.left, min, node.val - 1) &&
                isValidBST(node.right, node.val + 1, max);
    }

    public static void main(String[] args){

        TreeNode root = new TreeNode(-2147483648);
        root.left = new TreeNode(-2147483648);

        System.out.println((new Solution2()).isValidBST(root));
    }
}
